pkg/database/resync.go

```diff
@@ -4,45 +4,273 @@ package database
 	clusterName string, syncResponse *model.SyncResponse) error {
 
-	defer metrics.SlowLog(fmt.Sprintf("Slow Resync from cluster %s", clusterName), 0)()
+	defer metrics.SlowLog(fmt.Sprintf("Slow resync from %12s. RequestId: %d", clusterName, event.RequestId), 0)()
 	klog.Infof(
-		"Starting Resync of cluster %s. This is normal, but it could be a problem if it happens often.",
-		clusterName)
+		"Starting resync from %12s. This is normal, but it could be a problem if it happens often.", clusterName)
 
-	// DELETE from search.resources WHERE cluster=$1
-	delResourcesSql, args, delResourcesSqlErr := goquDelete("resources", "cluster", clusterName)
-	checkError(delResourcesSqlErr, fmt.Sprintf("Error creating query to delete cluster resources for %s.",
-		clusterName))
-
-	_, err := dao.pool.Exec(ctx, delResourcesSql, args...)
+	// Reset resources
+	err := dao.resetResources(ctx, event.AddResources, clusterName, syncResponse)
 	if err != nil {
-		klog.Warningf("Error deleting resources during resync of cluster %s. Error: %+v", clusterName, err)
+		klog.Warningf("Error resyncing resources for cluster %12s. Error: %+v", clusterName, err)
 		return err
 	}
-	// DELETE from search.edges WHERE cluster=$1
-	delEdgesSql, args, delEdgesSqlErr := goquDelete("edges", "cluster", clusterName)
-	checkError(delEdgesSqlErr, fmt.Sprintf("Error creating query to delete edges for %s.", clusterName))
 
-	_, err = dao.pool.Exec(ctx, delEdgesSql, args...)
+	// Reset edges
+	err = dao.resetEdges(ctx, event.AddEdges, clusterName, syncResponse)
 	if err != nil {
-		klog.Warningf("Error deleting edges during resync of cluster %s. Error: %+v", clusterName, err)
+		klog.Warningf("Error resyncing edges for cluster %12s. Error: %+v", clusterName, err)
 		return err
 	}
-	err = dao.SyncData(ctx, event, clusterName, syncResponse)
 
-	klog.V(1).Infof("Completed resync of cluster %s", clusterName)
-	return err
+	klog.V(1).Infof("Completed resync of cluster %12s.\t RequestId: %d", clusterName, event.RequestId)
+	return nil
+}
+
+// Reset Resources.
+//  1. Create a map of incoming resources.
+//  2. Query and iterate existing resources for the cluster.
+//  3. For each existing resource:
+//      - UPDATE if doesn't match the incoming resource.
+//      - DELETE if not found in the incoming resource.
+//  4. INSERT incoming resources not found in the existing resources.
+func (dao *DAO) resetResources(ctx context.Context, resources []model.Resource, clusterName string,
+	syncResponse *model.SyncResponse) error {
+	timer := time.Now()
+
+	batch := NewBatchWithRetry(ctx, dao, syncResponse)
+
+	incomingResMap := make(map[string]*model.Resource)
+	for i, resource := range resources {
+		incomingResMap[resource.UID] = &resources[i]
+	}
+	resourcesToDelete := make([]interface{}, 0)
+	resourcesToUpdate := make([]*model.Resource, 0)
+
+	// Get existing resources (UID and data) for the cluster.
+	query, params, err := useGoqu(
+		"SELECT uid, data FROM search.resources WHERE cluster=$1",
```

Comment:
Should we exclude the cluster node so that it doesn't get deleted?

Reply:
Updated query to 
SELECT uid, data FROM search.resources WHERE cluster=$1 AND uid!='cluster__$1'